File: Plugins/AM_ResultsCleanup_PlugIn/clsAnalysisToolRunnerResultsCleanup.vb
'*********************************************************************************************************
' Written by Matthew Monroe for the US Department of Energy 
' Pacific Northwest National Laboratory, Richland, WA
' Created 04/17/2013
'
'*********************************************************************************************************

Imports AnalysisManagerBase
Imports System.IO

Public Class clsAnalysisToolRunnerResultsCleanup
	Inherits clsAnalysisToolRunnerBase

#Region "Constants"
	Protected Const RESULTS_DB3_FILE As String = "Results.db3"
#End Region

#Region "Methods"

	''' <summary>
	''' Runs ResultsCleanup tool
	''' </summary>
	''' <returns>CloseOutType enum indicating success or failure</returns>
	''' <remarks></remarks>
	Public Overrides Function RunTool() As CloseOutType

		Try
			'Call base class for initial setup
			If Not MyBase.RunTool = CloseOutType.CLOSEOUT_SUCCESS Then
				Return CloseOutType.CLOSEOUT_FAILED
			End If

			' Store the AnalysisManager version info in the database
			If Not StoreToolVersionInfo() Then
				clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, "Aborting since StoreToolVersionInfo returned false")
				m_message = "Error determining AnalysisManager version"
				Return CloseOutType.CLOSEOUT_FAILED
			End If

			' Cleanup results in the transfer directory
			Dim Result As CloseOutType
			Result = PerformResultsCleanup()
			If Result <> CloseOutType.CLOSEOUT_SUCCESS Then
				If String.IsNullOrEmpty(m_message) Then
					m_message = "Unknown error calling PerformResultsCleanup"
				End If
				Return CloseOutType.CLOSEOUT_FAILED
			End If

			'Stop the job timer
			m_StopTime = System.DateTime.UtcNow

		Catch ex As Exception
			m_message = "Error in clsAnalysisToolRunnerResultsCleanup->RunTool"
			clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message & ": " & ex.Message)
			Return CloseOutType.CLOSEOUT_FAILED
		End Try

		'If we got to here, everything worked, so exit
		Return CloseOutType.CLOSEOUT_SUCCESS

	End Function

	Protected Function PerformResultsCleanup() As CloseOutType

		Dim strTransferDirectoryPath As String
		Dim strResultsFolderName As String
		Dim eResult As CloseOutType = CloseOutType.CLOSEOUT_SUCCESS

		Try

			strTransferDirectoryPath = m_jobParams.GetJobParameter("JobParameters", "transferFolderPath", String.Empty)
			strResultsFolderName = m_jobParams.GetJobParameter("JobParameters", "InputFolderName", String.Empty)

			If String.IsNullOrWhiteSpace(strTransferDirectoryPath) Then
				m_message = "transferFolderPath not defined"
				clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message)
				Return CloseOutType.CLOSEOUT_FAILED
			ElseIf String.IsNullOrWhiteSpace(strResultsFolderName) Then
				m_message = "InputFolderName not defined"
				clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message)
				Return CloseOutType.CLOSEOUT_FAILED
			End If

			Dim diTransferFolder As IO.DirectoryInfo = New IO.DirectoryInfo(strTransferDirectoryPath)

			If Not diTransferFolder.Exists Then
				m_message = "transferFolder not found at " & strTransferDirectoryPath
				clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message)
				Return CloseOutType.CLOSEOUT_FAILED
			End If

			Dim diResultsFolder As IO.DirectoryInfo
			diResultsFolder = New IO.DirectoryInfo(IO.Path.Combine(diTransferFolder.FullName, strResultsFolderName))
			eResult = RemoveOldResultsDb3Files(diResultsFolder)

		Catch ex As Exception
			m_message = "Error in PerformResultsCleanup"
			clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message & ": " & ex.Message)
			Return CloseOutType.CLOSEOUT_FAILED
		End Try

		Return eResult

	End Function

	Protected Function RemoveOldResultsDb3Files(ByVal diResultsFolder As IO.DirectoryInfo) As CloseOutType

		Dim reStepNumber As Text.RegularExpressions.Regex
		Dim reMatch As Text.RegularExpressions.Match

		Dim intStepFolderCount As Integer = 0
		Dim dctResultsFiles As Generic.Dictionary(Of Integer, IO.FileInfo)
		Dim intStepNumber As Integer

		Try
			reStepNumber = New Text.RegularExpressions.Regex("Step_(\d+)", Text.RegularExpressions.RegexOptions.Compiled Or Text.RegularExpressions.RegexOptions.IgnoreCase)
			dctResultsFiles = New Generic.Dictionary(Of Integer, IO.FileInfo)

			' Look for Results.db3 files in the subfolders of the transfer folder
			' Only process folders that start with the text "Step_"
			For Each diSubfolder In diResultsFolder.GetDirectories("Step_*")
				intStepFolderCount += 1

				' Parse out the step number
				reMatch = reStepNumber.Match(diSubfolder.Name)

				If reMatch.Success AndAlso Integer.TryParse(reMatch.Groups(1).Value, intStepNumber) Then
					If Not dctResultsFiles.ContainsKey(intStepNumber) Then
						For Each fiFile As IO.FileInfo In diSubfolder.GetFiles(RESULTS_DB3_FILE)
							dctResultsFiles.Add(intStepNumber, fiFile)
							Exit For
						Next
					End If
				End If
			Next

			If dctResultsFiles.Count > 1 Then
				' Delete the Results.db3 files for the steps prior to intLastStep
				Dim intLastStep As Integer = dctResultsFiles.Keys.Max
				Dim intFileCountDeleted As Integer = 0

				Dim lnqQuery = From item In dctResultsFiles Where item.Key < intLastStep
				For Each item In lnqQuery
					Try
						item.Value.Delete()
						intFileCountDeleted += 1
					Catch ex As Exception
						clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.WARN, "Error deleting extra " & RESULTS_DB3_FILE & " file: " & ex.Message)
					End Try
				Next

				m_EvalMessage = "Deleted " & intFileCountDeleted & " extra " & RESULTS_DB3_FILE & " " & clsGlobal.CheckPlural(intFileCountDeleted, "file", "files")
				clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.INFO, m_EvalMessage & " from " & diResultsFolder.FullName)

			ElseIf dctResultsFiles.Count = 1 Then
				m_EvalMessage = "Results folder has just one " & RESULTS_DB3_FILE & " file"
				clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.INFO, m_EvalMessage)
			Else
				If intStepFolderCount > 0 Then
					m_EvalMessage = "None of the Step_# folders has a " & RESULTS_DB3_FILE & " file"
					clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.WARN, m_EvalMessage)
				Else
					m_message = "Results folder does not have any Step_# folders"
					clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message & ": " & diResultsFolder.FullName)
					Return CloseOutType.CLOSEOUT_FAILED
				End If
			End If

		Catch ex As Exception
			m_message = "Error in RemoveOldResultsDb3Files"
			clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message & ": " & ex.Message)
			Return CloseOutType.CLOSEOUT_FAILED
		End Try

		Return CloseOutType.CLOSEOUT_SUCCESS

	End Function
	''' <summary>
	''' Stores the tool version info in the database
	''' </summary>
	''' <remarks></remarks>
	Protected Function StoreToolVersionInfo() As Boolean

		Dim strToolVersionInfo As String = String.Empty
		Dim strAppFolderPath As String = clsGlobal.GetAppFolderPath()

		If m_DebugLevel >= 2 Then
			clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.DEBUG, "Determining tool version info")
		End If

		' Lookup the version of the Analysis Manager
		If Not StoreToolVersionInfoForLoadedAssembly(strToolVersionInfo, "AnalysisManagerProg") Then
			Return False
		End If

		' Lookup the version of AnalysisManagerResultsCleanupPlugin
		If Not StoreToolVersionInfoForLoadedAssembly(strToolVersionInfo, "AnalysisManagerResultsCleanupPlugin") Then
			Return False
		End If

		' Store the path to AnalysisManagerProg.exe and AnalysisManagerResultsCleanupPlugin.dll in ioToolFiles
		Dim ioToolFiles As New List(Of FileInfo)
		ioToolFiles.Add(New FileInfo(Path.Combine(strAppFolderPath, "AnalysisManagerProg.exe")))
		ioToolFiles.Add(New FileInfo(Path.Combine(strAppFolderPath, "AnalysisManagerResultsCleanupPlugin.dll")))

		Try
			Return MyBase.SetStepTaskToolVersion(strToolVersionInfo, ioToolFiles, blnSaveToolVersionTextFile:=False)
		Catch ex As Exception
			clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, "Exception calling SetStepTaskToolVersion: " & ex.Message)
			Return False
		End Try

	End Function

#End Region
End Class

File: Plugins/AM_ResultsCleanup_PlugIn/clsAnalysisToolRunnerResultsCleanup.cs
//*********************************************************************************************************
// Written by Matthew Monroe for the US Department of Energy
// Pacific Northwest National Laboratory, Richland, WA
// Created 04/17/2013
//
//*********************************************************************************************************

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using AnalysisManagerBase;

namespace AnalysisManagerResultsCleanupPlugin
{
    /// <summary>
    /// Class for running Results Cleanup
    /// </summary>
    public class clsAnalysisToolRunnerResultsCleanup : clsAnalysisToolRunnerBase
    {
        #region "Constants"

        private const string RESULTS_DB3_FILE = "Results.db3";

        #endregion

        #region "Methods"

        /// <summary>
        /// Runs ResultsCleanup tool
        /// </summary>
        /// <returns>CloseOutType enum indicating success or failure</returns>
        public override CloseOutType RunTool()
        {
            try
            {
                // Call base class for initial setup
                if (base.RunTool() != CloseOutType.CLOSEOUT_SUCCESS)
                {
                    return CloseOutType.CLOSEOUT_FAILED;
                }

                // Store the AnalysisManager version info in the database
                if (!StoreToolVersionInfo())
                {
                    LogError("Aborting since StoreToolVersionInfo returned false");
                    mMessage = "Error determining AnalysisManager version";
                    return CloseOutType.CLOSEOUT_FAILED;
                }

                // Cleanup results in the transfer directory
                var Result = PerformResultsCleanup();
                if (Result != CloseOutType.CLOSEOUT_SUCCESS)
                {
                    if (string.IsNullOrEmpty(mMessage))
                    {
                        mMessage = "Unknown error calling PerformResultsCleanup";
                    }
                    return CloseOutType.CLOSEOUT_FAILED;
                }

                // Stop the job timer
                mStopTime = DateTime.UtcNow;
            }
            catch (Exception ex)
            {
                mMessage = "Error in clsAnalysisToolRunnerResultsCleanup->RunTool";
                LogError(mMessage + ": " + ex.Message);
                return CloseOutType.CLOSEOUT_FAILED;
            }

            // If we got to here, everything worked, so exit
            return CloseOutType.CLOSEOUT_SUCCESS;
        }

        private CloseOutType PerformResultsCleanup()
        {
            CloseOutType eResult;

            try
            {
                var strTransferDirectoryPath = mJobParams.GetJobParameter(clsAnalysisJob.JOB_PARAMETERS_SECTION, clsAnalysisResources.JOB_PARAM_TRANSFER_FOLDER_PATH, string.Empty);
                var strResultsFolderName = mJobParams.GetJobParameter(clsAnalysisJob.JOB_PARAMETERS_SECTION, "InputFolderName", string.Empty);

                if (string.IsNullOrWhiteSpace(strTransferDirectoryPath))
                {
                    mMessage = "transferFolderPath not defined";
                    LogError(mMessage);
                    return CloseOutType.CLOSEOUT_FAILED;
                }

                if (string.IsNullOrWhiteSpace(strResultsFolderName))
                {
                    mMessage = "InputFolderName not defined";
                    LogError(mMessage);
                    return CloseOutType.CLOSEOUT_FAILED;
                }

                var diTransferFolder = new DirectoryInfo(strTransferDirectoryPath);

                if (!diTransferFolder.Exists)
                {
                    mMessage = "transferFolder not found at " + strTransferDirectoryPath;
                    LogError(mMessage);
                    return CloseOutType.CLOSEOUT_FAILED;
                }

                var diResultsFolder = new DirectoryInfo(Path.Combine(diTransferFolder.FullName, strResultsFolderName));
                eResult = RemoveOldResultsDb3Files(diResultsFolder);
            }
            catch (Exception ex)
            {
                mMessage = "Error in PerformResultsCleanup";
                LogError(mMessage + ": " + ex.Message);
                return CloseOutType.CLOSEOUT_FAILED;
            }

            return eResult;
        }

        private CloseOutType RemoveOldResultsDb3Files(DirectoryInfo diResultsFolder)
        {
            var intStepFolderCount = 0;

            try
            {
                var reStepNumber = new Regex(@"Step_(\d+)", RegexOptions.Compiled | RegexOptions.IgnoreCase);
                var dctResultsFiles = new Dictionary<int, FileInfo>();

                // Look for Results.db3 files in the subfolders of the transfer folder
                // Only process folders that start with the text "Step_"
                foreach (var diSubfolder in diResultsFolder.GetDirectories("Step_*"))
                {
                    intStepFolderCount += 1;

                    // Parse out the step number
                    var reMatch = reStepNumber.Match(diSubfolder.Name);

                    if (reMatch.Success && int.TryParse(reMatch.Groups[1].Value, out var intStepNumber))
                    {
                        if (!dctResultsFiles.ContainsKey(intStepNumber))
                        {
                            foreach (var fiFile in diSubfolder.GetFiles(RESULTS_DB3_FILE))
                            {
                                dctResultsFiles.Add(intStepNumber, fiFile);
                                break;
                            }
                        }
                    }
                }

                if (dctResultsFiles.Count > 1)
                {
                    // Delete the Results.db3 files for the steps prior to intLastStep
                    var intLastStep = dctResultsFiles.Keys.Max();
                    var intFileCountDeleted = 0;

                    var lnqQuery = from item in dctResultsFiles where item.Key < intLastStep select item;
                    foreach (var item in lnqQuery)
                    {
                        try
                        {
                            item.Value.Delete();
                            intFileCountDeleted += 1;
                        }
                        catch (Exception ex)
                        {
                            LogWarning("Error deleting extra " + RESULTS_DB3_FILE + " file: " + ex.Message);
                        }
                    }

                    mEvalMessage = "Deleted " + intFileCountDeleted + " extra " + RESULTS_DB3_FILE + " " +
                                    clsGlobal.CheckPlural(intFileCountDeleted, "file", "files");

                    LogMessage(mEvalMessage + " from " + diResultsFolder.FullName);
                }
                else if (dctResultsFiles.Count == 1)
                {
                    mEvalMessage = "Results folder has just one " + RESULTS_DB3_FILE + " file";
                    LogMessage(mEvalMessage);
                }
                else
                {
                    if (intStepFolderCount > 0)
                    {
                        LogWarning("None of the Step_# folders has a " + RESULTS_DB3_FILE + " file", true);
                    }
                    else
                    {
                        mMessage = "Results folder does not have any Step_# folders";
                        LogError(mMessage + ": " + diResultsFolder.FullName);
                        return CloseOutType.CLOSEOUT_FAILED;
                    }
                }
            }
            catch (Exception ex)
            {
                mMessage = "Error in RemoveOldResultsDb3Files";
                LogError(mMessage + ": " + ex.Message);
                return CloseOutType.CLOSEOUT_FAILED;
            }

            return CloseOutType.CLOSEOUT_SUCCESS;
        }

        /// <summary>
        /// Stores the tool version info in the database
        /// </summary>
        /// <remarks></remarks>
        protected bool StoreToolVersionInfo()
        {
            var strToolVersionInfo = string.Empty;
            var strAppFolderPath = clsGlobal.GetAppDirectoryPath();

            if (mDebugLevel >= 2)
            {
                LogDebug("Determining tool version info");
            }

            // Lookup the version of the Analysis Manager
            if (!StoreToolVersionInfoForLoadedAssembly(ref strToolVersionInfo, "AnalysisManagerProg"))
            {
                return false;
            }

            // Lookup the version of AnalysisManagerResultsCleanupPlugin
            if (!StoreToolVersionInfoForLoadedAssembly(ref strToolVersionInfo, "AnalysisManagerResultsCleanupPlugin"))
            {
                return false;
            }

            // Store the path to AnalysisManagerProg.exe and AnalysisManagerResultsCleanupPlugin.dll in toolFiles
            var toolFiles = new List<FileInfo>
            {
                new FileInfo(Path.Combine(strAppFolderPath, "AnalysisManagerProg.exe")),
                new FileInfo(Path.Combine(strAppFolderPath, "AnalysisManagerResultsCleanupPlugin.dll"))
            };

            try
            {
                return SetStepTaskToolVersion(strToolVersionInfo, toolFiles, saveToolVersionTextFile: false);
            }
            catch (Exception ex)
            {
                LogError("Exception calling SetStepTaskToolVersion: " + ex.Message);
                return false;
            }
        }

        #endregion
    }
}
